massa-pos-exports/src/pos_final_state.rs

```diff
@@ -53,17 +58,97 @@ impl PoSFinalState {
         let init_seed = Hash::compute_from(initial_seed_string.as_bytes());
         let initial_seeds = vec![Hash::compute_from(init_seed.to_bytes()), init_seed];
 
+        let end_slot = Slot::new(0, config.thread_count.saturating_sub(1));
+
         Ok(Self {
-            config,
+            config: config.clone(),
             cycle_history: Default::default(),
             deferred_credits: DeferredCredits::default(),
             selector,
             initial_rolls,
             initial_seeds,
             initial_ledger_hash,
+            initial_cycle: end_slot.get_cycle(config.periods_per_cycle),
+            last_start_period: end_slot.period,
+        })
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    /// create a `PoSFinalState` from an existing snapshot
+    pub fn from_snapshot(
+        config: PoSConfig,
+        cycle_history: VecDeque<CycleInfo>,
+        deferred_credits: DeferredCredits,
+        initial_seed_string: &str,
+        initial_rolls_in_snapshot: BTreeMap<Address, u64>,
+        selector: Box<dyn SelectorController>,
+        initial_ledger_hash: Hash,
+        end_slot: Slot,
+    ) -> Result<Self, PosError> {
+        // Seeds used as the initial seeds for negative cycles (initial_cycle-2 and initial_cycle-1 respectively)
+        let init_seed = Hash::compute_from(initial_seed_string.as_bytes());
+        let initial_seeds = vec![Hash::compute_from(init_seed.to_bytes()), init_seed];
+
+        Ok(Self {
+            config: config.clone(),
+            cycle_history,
+            deferred_credits,
+            selector,
+            initial_rolls: initial_rolls_in_snapshot,
+            initial_seeds,
+            initial_ledger_hash,
+            initial_cycle: end_slot.get_cycle(config.periods_per_cycle),
+            last_start_period: end_slot.period,
         })
     }
 
+    /// Handles the deferred credits that were due during a downtime
+    pub fn update_deferred_credits_after_restart(
+        &mut self,
+        from_slot: Slot,
+        end_slot: Slot,
+    ) -> Result<(), PosError> {
+        let mut deferred_credits = self.deferred_credits.clone();
+
+        let mut updated_deferred_credits = DeferredCredits::default();
+        let next_slot = end_slot
+            .get_next_slot(self.config.thread_count)
+            .map_err(|_| {
+                PosError::OverflowError(String::from("Cannot get next slot after genesis"))
+            })?;
+
+        for (slot, map) in deferred_credits.credits.clone() {
```

Comment:
```rust
let iter = deferred_credits.credits.clone()
    .iter()
    .filter(|(slot, _)| slot > from_slot) // even better: if this is known to be sorted, do a binary search
    .take_while(|(slot, _)| slot <= end_slot);
for (slot, map) in iter {
    for ... in map.into_iter() { 
        ....
    }
}
```

Then look into using flatten so that you only have one `for`, which goes straight into the the iterator over all (addr, amount)s

Reply:
Used `BTreeMap`'s `range` method to iterate in da77c706d49425ce526501a0a2facdcc923bf4df.
However, I still need the slot variable in the inner loop, so I couldn't make the flatten idea work. Maybe you have an idea?

Reply:
BtreeMap iteration would probably be less performant than a sort => binary-search => take-until: collecting a btreemap is O(n log_n), same as the sort, but iterating through a btreemap is not nearly as clean as iterating through a vector (tree-traversal and cache misses), but this is only relevant if this is a performance hot-spot.

The issue of readability is what prompted this comment initially: The filter and take_while methods will handle the need for the `continue` and `break` checks, and the thinking was the `flatten` to remove the nested `for`.

using the filter/bin-search + take_while should be easy enough, and make the code easier to parse for the next person to come along. The `flatten` benefit to readability is debatable, though.